primitives/src/market.rs

```diff
@@ -125,8 +125,8 @@ pub struct MarketDispute<AccountId, BlockNumber> {
 
 /// How a market should resolve disputes
 #[derive(Clone, Decode, Encode, Eq, MaxEncodedLen, PartialEq, RuntimeDebug, TypeInfo)]
-pub enum MarketDisputeMechanism<AI> {
-    Authorized(AI),
```

Comment:
what happens for old market data which has account_Id in authorized variant of MDM and we try to decode them with this new definition, to me it seems we need migration @maltekliemann please advise :pray:

Reply:
Yeah, we should do a storage migration.

The scalar codec encodes instances of enums using one byte to designate which variant is instance is, followed by an encoding of the instances that the variant contains. For example,

```rust
enum OldEnum {
    A,
    B(u8),
}
```

encodes `OldEnum::A` to `0x00` and `OldEnum::B(2)` to `0x0102`.

Furthermore,

> For structures, the values are named, but that is irrelevant for the encoding (names are ignored - only order matters). All containers store elements consecutively. The order of the elements is not fixed, depends on the container, and cannot be relied on at decoding. This implicitly means that decoding some byte-array into a specified structure that enforces an order and then re-encoding it could result in a different byte array than the original that was decoded.

So if you have a struct which contains a field of type `OldEnum` and now replace `OldEnum` with...

```rust
enum NewEnum {
    A,
    B,
}
```

... then its encoding in scale will be a byte shorter. So if you put an `OldMarket` into the decoder of the `NewMarket`, then the scale codec will either reject the byte sequence on the grounds of it being too long, or it will ignore the last byte, giving you a faulty decoding.

I don't know what exactly the scale codec will do in this situation, but I don't intend to find out. However, you can try this with `py-scale-codec` and get a big fat error:

```python
from scalecodec.base import RuntimeConfiguration
from scalecodec.type_registry import load_type_registry_preset

RuntimeConfiguration().update_type_registry(load_type_registry_preset("default"))

my_types = {
    "types": {
        "OldEnum": {
            "type": "enum",
            "type_mapping": [
                ("A", "null"),
                ("B", "u8"),
            ],
        },
        "OldStruct": {
            "type": "struct",
            "type_mapping": [
                ("enum", "OldEnum"),
                ("other", "u8"),
            ],
        },
        "NewEnum": {
            "type": "enum",
            "type_mapping": [
                ("A", "null"),
                ("B", "null"),
            ],
        },
        "NewStruct": {
            "type": "struct",
            "type_mapping": [
                ("enum", "NewEnum"),
                ("other", "u8"),
            ],
        },
    }
}

RuntimeConfiguration().update_type_registry(my_types)

old_struct = RuntimeConfiguration().create_scale_object("OldStruct")
old_raw = old_struct.encode({"enum": {"B": 2}, "other": 3})
print("old_struct: ", old_raw)

new_struct = RuntimeConfiguration().create_scale_object("NewStruct")
new_raw = new_struct.encode({"enum": "B", "other": 3})
print("new_struct: ", new_raw)

new_struct_from_old_raw = RuntimeConfiguration().create_scale_object(
    "NewStruct", old_raw
)
# This raises a scalecodec.exceptions.RemainingScaleBytesNotEmptyException
decoded = new_struct_from_old_raw.decode()
```